{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2c592c87",
   "metadata": {},
   "source": [
    "# Assigment 2 - Python dictionaries and file I/O\n",
    "\n",
    "### Overview\n",
    "------------\n",
    "\n",
    "The GTEx gene expression dataset consists of samples taken from many human subjects across a variety of tissues. From this dataset we have identified a set of highly-expressed tissue-specific genes. In this exercise, you will be pulling the expressing data for those genes but only in their primary tissue of expression and investigating how much variability exists across the sample population.\n",
    "\n",
    "Biological Learning Objectives\n",
    "\n",
    "- Use metadata to associate data with subject or sample traits\n",
    "- Use descriptive statistics to make inferences about gene expression heterogeneity\n",
    "\n",
    "Computational Learning Objectives\n",
    "\n",
    "- Create and modify a dicitionary\n",
    "- Use keys to retrieve values stored in a dictionary\n",
    "- Step through each key and value in a dictionary inside a `for` loop\n",
    "- Open and read a text file\n",
    "- Write to a text file\n",
    "\n",
    "### Instructions\n",
    "----------------\n",
    "\n",
    "- Save a copy of this notebook as `~/qbXX-answers/python_dictionaries.ipynb`.\n",
    "- Fill in answers in the available code/markdown cells below.\n",
    "- Remember to comment your code to help yourself and us know what each part is intended to do."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ba1d85fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use this dictionary for identifying genes of interest and their corresponding tissue of interest\n",
    "\n",
    "gene_tissue = {\n",
    "    'ENSG00000042832.11': 'Thyroid',\n",
    "    'ENSG00000091704.9': 'Pancreas',\n",
    "    'ENSG00000118245.2': 'Testis',\n",
    "    'ENSG00000122304.10': 'Testis',\n",
    "    'ENSG00000122852.14': 'Lung',\n",
    "    'ENSG00000132693.12': 'Liver',\n",
    "    'ENSG00000134812.7': 'Stomach',\n",
    "    'ENSG00000135346.8': 'Pituitary',\n",
    "    'ENSG00000137392.9': 'Pancreas',\n",
    "    'ENSG00000142515.14': 'Prostate',\n",
    "    'ENSG00000142615.7': 'Pancreas',\n",
    "    'ENSG00000142789.19': 'Pancreas',\n",
    "    'ENSG00000158874.11': 'Liver',\n",
    "    'ENSG00000162438.11': 'Pancreas',\n",
    "    'ENSG00000164816.7': 'Small Intestine - Terminal Ileum',\n",
    "    'ENSG00000164822.4': 'Small Intestine - Terminal Ileum',\n",
    "    'ENSG00000168925.10': 'Pancreas',\n",
    "    'ENSG00000168928.12': 'Pancreas',\n",
    "    'ENSG00000170890.13': 'Pancreas',\n",
    "    'ENSG00000171195.10': 'Minor Salivary Gland',\n",
    "    'ENSG00000172179.11': 'Pituitary',\n",
    "    'ENSG00000175535.6': 'Pancreas',\n",
    "    'ENSG00000175646.3': 'Testis',\n",
    "    'ENSG00000182333.14': 'Stomach',\n",
    "    'ENSG00000187021.14': 'Pancreas',\n",
    "    'ENSG00000203784.2': 'Testis',\n",
    "    'ENSG00000204983.13': 'Pancreas',\n",
    "    'ENSG00000219073.7': 'Pancreas',\n",
    "    'ENSG00000229859.9': 'Stomach',\n",
    "    'ENSG00000240338.5': 'Pancreas',\n",
    "    'ENSG00000254647.6': 'Pancreas',\n",
    "    'ENSG00000256713.7': 'Stomach',\n",
    "    'ENSG00000259384.6': 'Pituitary',\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b9449c2",
   "metadata": {},
   "source": [
    "1. Load sample metadata from \"~/Data/GTEx/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt\" and create a dictionary of sample metadata associating each sample ID with the tissue it was sampled from\n",
    "\n",
    "    - This file is tab-separated but some values have spaces in their text so make sure to specify `\\t` in the `.split()` method\n",
    "    - There are several columns in the metadata file, but you only need to be concerned with `SAMPID` and `SMTSD`, the sample ID and sample tissue, respectively\n",
    "    - Your dictionary should use the `SAMPID` index [0] as the key and `SMTSD` as the value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8f7622d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_fname = \"/Users/cmdb/Data/GTEx/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt\"\n",
    "md_file = open(metadata_fname)\n",
    "meta_dictionary = {} # Create a dictionary to hold the metadata\n",
    "header= md_file.readline() # the readline reads until the first newline character, use this an print it such that you can find the index positions for the columns you are interested in\n",
    "for line in md_file: # Step through each line of the metadata file, stepping through a line converts to a list\n",
    "    if line == header:\n",
    "        continue\n",
    "    line = line.strip('\\n')\n",
    "    line = line.split('\\t') # Remove the newline character from the line you just read in and split it by the tab character\n",
    "    meta_dictionary[line[0]] = line[6]  # the dictionary has to be in the forloop because you are adding the key for each line, refer to the fruit slide\n",
    "    # Add the data to your metadata dictionary, using the \"SAMPID\" column value as the dictionary key and the \"SMTSD\" column value as the dictionary value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1821cf32",
   "metadata": {},
   "source": [
    "2. Load and process the headeer from the GTEx gene expression file \"~/Data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz\"\n",
    "\n",
    "    - You have code that will open the file and read in the header line\n",
    "    - Create a list of tissues corresponding to the column names in the header, one tissue per column name (don't worry about which the column is a sample ID or not since column names that aren't sample IDs will just get a \"missing\" label)\n",
    "    - For each value in the header, see if it appears in the dictionary of sample ID/tissues that you created in step 1\n",
    "    - If the value is in your dictionary, add the corresponding tissue name (the value in dictionary for that sample ID key) to your tissue list\n",
    "    - If the value is not in your dictionary, add \"missing\" to the tissue list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "544a432c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Because this file is a gzip-compressed file, you will be using the `gzip` library that comes with python\n",
    "import gzip\n",
    "\n",
    "expression_fname = \"/Users/cmdb/Data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz\"\n",
    "\n",
    "# Open the file using the gzip library\n",
    "fs = gzip.open(expression_fname)\n",
    "\n",
    "# The file starts with two lines of information before the header, so these are skipped using 2 calls of `.readline()` without keeping the returned data\n",
    "_ = fs.readline()\n",
    "_ = fs.readline()\n",
    "\n",
    "# Gzipped text files are read in as byte strings, not regular strings, so you will see that `.decode()` is included after `.readline()` to convert the input into a string\n",
    "header = fs.readline().decode().rstrip().split(\"\\t\")\n",
    "\n",
    "# Make sure to close the file when you are done reading it\n",
    "fs.close()\n",
    "\n",
    "# Create a list to hold the tissue names\n",
    "tissues = []\n",
    "\n",
    "# Step through each value in the header\n",
    "for value in header:\n",
    "    if value not in meta_dictionary:\n",
    "        tissues.append('Missing')\n",
    "    else:\n",
    "        tissues.append(meta_dictionary.get(value)) # you want to print the value onto this because it will take the individual index position and stick it onto the list. appending header would take that whole long list and append it every time\n",
    "# Check if the value has an entry in the gene_tissue dictionary, adding either the tissue corresponding to that entry or \"missing\" to your tissues list\n",
    "# Example of the start of the print(tissues) statement. ['Missing', 'Missing', 'Adipose - Subcutaneous', 'Muscle - Skeletal', 'Artery - Tibial'...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eed39f12",
   "metadata": {},
   "source": [
    "3. Create a list of expression values for each gene in the `gene_tissue` dictionary (given at the start of the exercise), keeping only expression values from the tissue of interest for that gene\n",
    "\n",
    "    - Since the file is already open, you can keep reading lines from it in, picking up from after the last line read\n",
    "    - You will be log-transforming the expression data using the `log2` function from the built-in library `math`\n",
    "    - To deal with zeros in expression data, add one before log-transforming expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5b44a45b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the log2 function from the math library to transform the expression counts\n",
    "from math import log2\n",
    "\n",
    "# Create a dictionary to hold each gene's expression data\n",
    "gene_expression = {}\n",
    "\n",
    "# Because this is a large file, for testing purposes you can use only a small portion of the file to test your code\n",
    "line_counter = 0\n",
    "\n",
    "# Step through each line of the file\n",
    "for each in gzip.open(expression_fname):\n",
    "\n",
    "    # Convert the line into a string from a byte string, remove the newline character at the end, and split it by tab characters\n",
    "    fields = each.decode().rstrip().split(\"\\t\")\n",
    "\n",
    "    # Check if this line contains data for a gene we are interested in. If not, skip it\n",
    "    fields = each.decode().rstrip().split(\"\\t\")\n",
    "    if fields[0] not in gene_tissue:\n",
    "        continue\n",
    "\n",
    "    # Determine which tissue we care about for this gene\n",
    "    tissue = gene_tissue[fields[0]]\n",
    "\n",
    "    # Create a list in the gene expression dictionary to hold the gene's data\n",
    "    gene_expression[fields[0]] = []\n",
    "\n",
    "    # Look at each column position, one by one\n",
    "    for i in range(len(tissues)):\n",
    "\n",
    "        # For each column, see if it is from the tissue we care about (using the tissues list). If so, add that field to our gene expression list, first transforming it with log2(1 + float(field[i])) \n",
    "        if tissues[i] == tissue:\n",
    "            gene_expression[fields[0]].append(log2(1 + float(fields[i])))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4676fc06",
   "metadata": {},
   "source": [
    "4. For each gene, calculate the mean, standard deviation, and relative standard deviation and write them to a results file\n",
    "\n",
    "    - You will be using the `mean` and `stdev` functions from the built-in `statistics` library. Both functions accept a list as their input\n",
    "    - Format your results for each gene into a string and write that string to your results file\n",
    "    - If you want to organize your results better (completely optional!!), you can save your results in a list and sorting them by tissue type before writing them to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6b89b161",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The tisue name is Pancreas, the gene name is ENSG00000162438.11, the mean is  11.64813927955005, the stdev is 0.757491204101109, the relative stdev is 0.06503109088255767\n",
      "The tisue name is Pancreas, the gene name is ENSG00000142615.7, the mean is  11.982260919355417, the stdev is 1.4113620914457623, the relative stdev is 0.1177876279731093\n",
      "The tisue name is Pancreas, the gene name is ENSG00000219073.7, the mean is  13.186023575042132, the stdev is 0.6691398656550582, the relative stdev is 0.05074614510181628\n",
      "The tisue name is Pancreas, the gene name is ENSG00000142789.19, the mean is  14.651792140427176, the stdev is 0.5222110990262671, the relative stdev is 0.03564144877440514\n",
      "The tisue name is Testis, the gene name is ENSG00000203784.2, the mean is  10.588661238478307, the stdev is 1.2485963275387792, the relative stdev is 0.11791824286544222\n",
      "The tisue name is Liver, the gene name is ENSG00000132693.12, the mean is  11.233820059075176, the stdev is 3.2303438267149205, the relative stdev is 0.2875552403125156\n",
      "The tisue name is Liver, the gene name is ENSG00000158874.11, the mean is  12.403165379794865, the stdev is 0.6799587560178442, the relative stdev is 0.054821389153249356\n",
      "The tisue name is Testis, the gene name is ENSG00000118245.2, the mean is  12.68051955583237, the stdev is 1.4374443015264093, the relative stdev is 0.11335847046308609\n",
      "The tisue name is Minor Salivary Gland, the gene name is ENSG00000171195.10, the mean is  13.66240542509939, the stdev is 3.508731365908907, the relative stdev is 0.2568165163261053\n",
      "The tisue name is Pituitary, the gene name is ENSG00000172179.11, the mean is  15.34070940366395, the stdev is 2.0224475699917153, the relative stdev is 0.13183533543166376\n",
      "The tisue name is Pancreas, the gene name is ENSG00000137392.9, the mean is  15.263872743216705, the stdev is 0.847495476085084, the relative stdev is 0.05552296526199176\n",
      "The tisue name is Pituitary, the gene name is ENSG00000135346.8, the mean is  11.211051271082567, the stdev is 2.05181509842288, the relative stdev is 0.18301718980763804\n",
      "The tisue name is Pancreas, the gene name is ENSG00000091704.9, the mean is  15.458753143882536, the stdev is 0.717376316049403, the relative stdev is 0.04640583295252949\n",
      "The tisue name is Pancreas, the gene name is ENSG00000204983.13, the mean is  16.137787479687645, the stdev is 0.7930013108179902, the relative stdev is 0.049139407233868164\n",
      "The tisue name is Small Intestine - Terminal Ileum, the gene name is ENSG00000164822.4, the mean is  10.555235965173559, the stdev is 3.1738956134018856, the relative stdev is 0.30069395169127305\n",
      "The tisue name is Small Intestine - Terminal Ileum, the gene name is ENSG00000164816.7, the mean is  11.725206247846716, the stdev is 3.0625750982778186, the relative stdev is 0.2611958402727668\n",
      "The tisue name is Thyroid, the gene name is ENSG00000042832.11, the mean is  12.68872987138559, the stdev is 0.9470126983439187, the relative stdev is 0.07463416023060992\n",
      "The tisue name is Lung, the gene name is ENSG00000122852.14, the mean is  10.094051754993973, the stdev is 2.900195718519805, the relative stdev is 0.2873173022007689\n",
      "The tisue name is Stomach, the gene name is ENSG00000182333.14, the mean is  12.506637140125969, the stdev is 5.011620164552823, the relative stdev is 0.4007168440566386\n",
      "The tisue name is Pancreas, the gene name is ENSG00000175535.6, the mean is  14.577667957161138, the stdev is 1.5774509350971981, the relative stdev is 0.10821010189920609\n",
      "The tisue name is Pancreas, the gene name is ENSG00000187021.14, the mean is  11.352044322367236, the stdev is 0.60805174093226, the relative stdev is 0.05356319299548535\n",
      "The tisue name is Pancreas, the gene name is ENSG00000254647.6, the mean is  10.997384243879184, the stdev is 1.8848077384973996, the relative stdev is 0.17138691316950458\n",
      "The tisue name is Stomach, the gene name is ENSG00000134812.7, the mean is  8.290498779007475, the stdev is 4.1522837635540215, the relative stdev is 0.5008484862295736\n",
      "The tisue name is Stomach, the gene name is ENSG00000229859.9, the mean is  8.732697313784572, the stdev is 4.595783632269953, the relative stdev is 0.5262730937685776\n",
      "The tisue name is Stomach, the gene name is ENSG00000256713.7, the mean is  9.42991153416666, the stdev is 4.491977041839647, the relative stdev is 0.4763541021105254\n",
      "The tisue name is Pancreas, the gene name is ENSG00000170890.13, the mean is  14.073429783545471, the stdev is 0.5720383460979774, the relative stdev is 0.04064669060038226\n",
      "The tisue name is Testis, the gene name is ENSG00000122304.10, the mean is  14.012950172919384, the stdev is 1.428699006557175, the relative stdev is 0.10195561883308456\n",
      "The tisue name is Testis, the gene name is ENSG00000175646.3, the mean is  13.921464122485672, the stdev is 1.498209066406657, the relative stdev is 0.10761864220781056\n",
      "The tisue name is Pancreas, the gene name is ENSG00000168928.12, the mean is  14.362048866146097, the stdev is 0.5788730638807795, the relative stdev is 0.04030574392803288\n",
      "The tisue name is Pancreas, the gene name is ENSG00000168925.10, the mean is  14.136953933949323, the stdev is 0.4183678441879013, the relative stdev is 0.02959391720045206\n",
      "The tisue name is Pancreas, the gene name is ENSG00000240338.5, the mean is  11.7016676968443, the stdev is 0.7054571422813039, the relative stdev is 0.06028688906210789\n",
      "The tisue name is Pituitary, the gene name is ENSG00000259384.6, the mean is  15.281443747697883, the stdev is 2.1283772961155414, the relative stdev is 0.13927854797333378\n",
      "The tisue name is Prostate, the gene name is ENSG00000142515.14, the mean is  11.359285573911217, the stdev is 2.7824494085167393, the relative stdev is 0.24494933157655346\n",
      "['Liver', 0.054821389153249356]\n",
      "['Liver', 0.2875552403125156]\n",
      "['Lung', 0.2873173022007689]\n",
      "['Minor Salivary Gland', 0.2568165163261053]\n",
      "['Pancreas', 0.02959391720045206]\n",
      "['Pancreas', 0.03564144877440514]\n",
      "['Pancreas', 0.04030574392803288]\n",
      "['Pancreas', 0.04064669060038226]\n",
      "['Pancreas', 0.04640583295252949]\n",
      "['Pancreas', 0.049139407233868164]\n",
      "['Pancreas', 0.05074614510181628]\n",
      "['Pancreas', 0.05356319299548535]\n",
      "['Pancreas', 0.05552296526199176]\n",
      "['Pancreas', 0.06028688906210789]\n",
      "['Pancreas', 0.06503109088255767]\n",
      "['Pancreas', 0.10821010189920609]\n",
      "['Pancreas', 0.1177876279731093]\n",
      "['Pancreas', 0.17138691316950458]\n",
      "['Pituitary', 0.13183533543166376]\n",
      "['Pituitary', 0.13927854797333378]\n",
      "['Pituitary', 0.18301718980763804]\n",
      "['Prostate', 0.24494933157655346]\n",
      "['Small Intestine - Terminal Ileum', 0.2611958402727668]\n",
      "['Small Intestine - Terminal Ileum', 0.30069395169127305]\n",
      "['Stomach', 0.4007168440566386]\n",
      "['Stomach', 0.4763541021105254]\n",
      "['Stomach', 0.5008484862295736]\n",
      "['Stomach', 0.5262730937685776]\n",
      "['Testis', 0.10195561883308456]\n",
      "['Testis', 0.10761864220781056]\n",
      "['Testis', 0.11335847046308609]\n",
      "['Testis', 0.11791824286544222]\n",
      "['Thyroid', 0.07463416023060992]\n"
     ]
    }
   ],
   "source": [
    "# Load the functions to caculcate the mean and standard deviation from a list\n",
    "from statistics import stdev, mean\n",
    "\n",
    "# Open a file to write the results to\n",
    "output = open('day2hw1.txt', 'w')\n",
    "\n",
    "# Step through each gene and its expression value list\n",
    "sorting = []\n",
    "for gene in gene_expression:\n",
    "    expression_list = gene_expression[gene] #access entries in adictionary using square brackets #access values by using name of dictionary, then the value you want to match to the key\n",
    "    #expression_list = gene_expression.get(gene)  # Calculate the mean and standard deviation of expression\n",
    "    gene_mean = mean(expression_list)\n",
    "    gene_stdev = stdev(expression_list)\n",
    "    gene_reldev= gene_stdev / gene_mean\n",
    "    tis = gene_tissue[gene] #so the point of this is to match the gene name to to tissue type. in part 4m you have the gene_expression in a dictionary so when you parse thru this dictionary, the index variable (gene) is defined by all of the gene expression key is taking on the values of interest from the gene_expression index. You can find the matching tissue using the gene variable which is the key from the gene_tissue dictionary\n",
    "    sorting.append([tis,gene_reldev]) #remember that you can append this as a list, by containing your variables as a bracket, this command is grouping two variables as a list, such that they are in a long list of grouped sublists\n",
    "    print (f'The tisue name is {tis}, the gene name is {gene}, the mean is  {gene_mean}, the stdev is {gene_stdev}, the relative stdev is {gene_reldev}')\n",
    "    output.write(f'The tisue name is {tis}, the gene name is {gene}, the mean is  {gene_mean}, the stdev is {gene_stdev}, the relative stdev is {gene_reldev}\\n')\n",
    "    # Calculate the relative standard deviation (coefficient of variation, std / mean)\n",
    "    \n",
    "    # Write the tissue namne, gene name, mean, standard deviation, and relative standard deviation to your results file\n",
    "\n",
    "# Close your results file\n",
    "output.close()\n",
    "sorting.sort() #sorting my list of just tissue name and relative std dev\n",
    "for one in sorting: #printing each sublist as an individual line to make the analysis much easier\n",
    "    print(one)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "185fe1ce",
   "metadata": {},
   "source": [
    "5. Looking at your results, answer the following questions\n",
    "\n",
    "    - Which tissues have low relative variability (rv < 0.2), medium relative variability (0.2 < rv < 0.4), or high relative variability (rv > 0.4)?\n",
    "    - Do each of the genes for a given tissue show the same level of relative variability?\n",
    "    - Propose an explanation for differences in relative variability you see between tissues based on what you know about their functions.\n",
    "    - Were any of the results surpising to you?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69c9d544",
   "metadata": {},
   "source": [
    "The tissues with low relative variability (rv<0.2) are the pancreas, pituitary, testis, and thyroid. The liver, lung, small intestine, minor salivary gland, prostate have medium relative variability (0.2< rv < 0.4). The stomach had a high relative variability (rv > 0.4). However, some of the tissue types- for example the liver- had differing relative variability depending on what genes were expressed, thus falling into the medium variability classification.\n",
    "\n",
    "It really depends on the tissue type. For example, the four genes in the testis tissue had a relative variability ranging between 0.1 and 0.11. This is very consistent juxtaposed to the spread of relative variability seen in the Liver; Where one gene has rv~0.05 and the other has rv~0.2.\n",
    "\n",
    "The differences in relative variability seen between the tissues could be a result of the gene-function and cell-type variability seen in the different tissue types. For example, one gene may be reserved for a specific function in a rare cell type, and therefore has a higher variability compared to a housekeeping gene which might be necessary for the identity of all cell types in the tissue.\n",
    "\n",
    "\n",
    "I was surprised by the high relative varability of genes seen in the Stomach. I would have assumed that with such a specialized function, there would be relatively stable expression of genes within this tissue. However, upon further reflection, I understand that the variety of substrates the stomach encounters must require specialized gene arrays for the most effective digestion. Depending on what substrates are present at the time, could be an effector for what genes get expressed, making it a highly malleable system."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a3f6e9a",
   "metadata": {},
   "source": [
    "Answers:\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qb25",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
